File: minigit-4/ts/src/commands/diff.ts
// diff command - Show changes

import * as fs from 'fs';
import * as path from 'path';
import { findRepoRoot, normalizePathSeparator, isBinaryFile } from '../utils';
import { readIndex } from '../index-file';
import { readObject, parseCommitContent, parseTreeContent, createBlobContent, hashObject, objectExists } from '../objects';
import { getHeadCommit, resolveRevision } from '../refs';
import { formatUnifiedDiff, formatDiffStat } from '../diff-algo';
import { FileDiff } from '../types';

export function diff(args: string[]): number {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a minigit repository');
    return 1;
  }

  let cached = false;
  let showStat = false;
  const commits: string[] = [];
  const paths: string[] = [];
  let inPathSpec = false;

  for (const arg of args) {
    if (arg === '--') {
      inPathSpec = true;
    } else if (inPathSpec) {
      paths.push(arg);
    } else if (arg === '--cached' || arg === '--staged') {
      cached = true;
    } else if (arg === '--stat') {
      showStat = true;
    } else if (!arg.startsWith('-')) {
      commits.push(arg);
    }
  }

  let output: string;

  if (commits.length === 2) {
    // Diff between two commits
    output = diffBetweenCommits(repoRoot, commits[0], commits[1], paths);
  } else if (commits.length === 1) {
    // Diff between commit and working tree
    output = diffCommitToWorkingTree(repoRoot, commits[0], paths);
  } else if (cached) {
    // Diff between HEAD and index (staged changes)
    output = diffHeadToIndex(repoRoot, paths);
  } else {
    // Diff between index and working tree (unstaged changes)
    output = diffIndexToWorkingTree(repoRoot, paths);
  }

  if (output) {
    console.log(output.trimEnd());
  }

  return 0;
}

function diffIndexToWorkingTree(repoRoot: string, filterPaths: string[]): string {
  const entries = readIndex(repoRoot);
  const diffs: string[] = [];

  for (const entry of entries) {
    if (filterPaths.length > 0 && !matchesPath(entry.name, filterPaths)) {
      continue;
    }

    const fullPath = path.join(repoRoot, entry.name);

    if (!fs.existsSync(fullPath)) {
      // File deleted
      const { content: oldContent } = readObject(repoRoot, entry.sha);
      if (isBinaryFile(oldContent)) {
        diffs.push(`Binary file ${entry.name} deleted`);
      } else {
        diffs.push(formatUnifiedDiff(entry.name, entry.name, oldContent.toString(), ''));
      }
      continue;
    }

    const stat = fs.lstatSync(fullPath);
    let workingContent: Buffer;

    if (stat.isSymbolicLink()) {
      workingContent = Buffer.from(fs.readlinkSync(fullPath));
    } else {
      workingContent = fs.readFileSync(fullPath);
    }

    const workingBlobContent = createBlobContent(workingContent);
    const workingSha = hashObject(workingBlobContent);

    if (workingSha !== entry.sha) {
      const { content: indexContent } = readObject(repoRoot, entry.sha);

      if (isBinaryFile(indexContent) || isBinaryFile(workingContent)) {
        diffs.push(`Binary files a/${entry.name} and b/${entry.name} differ`);
      } else {
        diffs.push(formatUnifiedDiff(entry.name, entry.name, indexContent.toString(), workingContent.toString()));
      }
    }
  }

  return diffs.join('');
}

function diffHeadToIndex(repoRoot: string, filterPaths: string[]): string {
  const entries = readIndex(repoRoot);
  const headCommit = getHeadCommit(repoRoot);

  // Get HEAD tree files
  const headFiles = new Map<string, string>();
  if (headCommit) {
    const { content } = readObject(repoRoot, headCommit);
    const commitInfo = parseCommitContent(content);
    collectTreeFiles(repoRoot, commitInfo.tree, '', headFiles);
  }

  const diffs: string[] = [];

  // Check index files
  for (const entry of entries) {
    if (filterPaths.length > 0 && !matchesPath(entry.name, filterPaths)) {
      continue;
    }

    const headSha = headFiles.get(entry.name);

    if (!headSha) {
      // New file
      const { content } = readObject(repoRoot, entry.sha);
      if (isBinaryFile(content)) {
        diffs.push(`Binary file ${entry.name} added`);
      } else {
        diffs.push(formatUnifiedDiff('/dev/null', entry.name, '', content.toString()));
      }
    } else if (headSha !== entry.sha) {
      // Modified
      const { content: headContent } = readObject(repoRoot, headSha);
      const { content: indexContent } = readObject(repoRoot, entry.sha);

      if (isBinaryFile(headContent) || isBinaryFile(indexContent)) {
        diffs.push(`Binary files a/${entry.name} and b/${entry.name} differ`);
      } else {
        diffs.push(formatUnifiedDiff(entry.name, entry.name, headContent.toString(), indexContent.toString()));
      }
    }

    headFiles.delete(entry.name);
  }

  // Check for deleted files (in HEAD but not in index)
  for (const [name, sha] of headFiles) {
    if (filterPaths.length > 0 && !matchesPath(name, filterPaths)) {
      continue;
    }

    const { content } = readObject(repoRoot, sha);
    if (isBinaryFile(content)) {
      diffs.push(`Binary file ${name} deleted`);
    } else {
      diffs.push(formatUnifiedDiff(name, '/dev/null', content.toString(), ''));
    }
  }

  return diffs.join('');
}

function diffBetweenCommits(repoRoot: string, ref1: string, ref2: string, filterPaths: string[]): string {
  const sha1 = resolveRevision(repoRoot, ref1);
  const sha2 = resolveRevision(repoRoot, ref2);

  if (!sha1 || !sha2) {
    return '';
  }

  const files1 = new Map<string, string>();
  const files2 = new Map<string, string>();

  const { content: content1 } = readObject(repoRoot, sha1);
  const commitInfo1 = parseCommitContent(content1);
  collectTreeFiles(repoRoot, commitInfo1.tree, '', files1);

  const { content: content2 } = readObject(repoRoot, sha2);
  const commitInfo2 = parseCommitContent(content2);
  collectTreeFiles(repoRoot, commitInfo2.tree, '', files2);

  return diffFileMaps(repoRoot, files1, files2, filterPaths);
}

function diffCommitToWorkingTree(repoRoot: string, ref: string, filterPaths: string[]): string {
  const sha = resolveRevision(repoRoot, ref);
  if (!sha) {
    return '';
  }

  const commitFiles = new Map<string, string>();
  const { content } = readObject(repoRoot, sha);
  const commitInfo = parseCommitContent(content);
  collectTreeFiles(repoRoot, commitInfo.tree, '', commitFiles);

  const diffs: string[] = [];

  // Compare commit files to working tree
  for (const [name, fileSha] of commitFiles) {
    if (filterPaths.length > 0 && !matchesPath(name, filterPaths)) {
      continue;
    }

    const fullPath = path.join(repoRoot, name);

    if (!fs.existsSync(fullPath)) {
      // Deleted
      const { content: fileContent } = readObject(repoRoot, fileSha);
      if (isBinaryFile(fileContent)) {
        diffs.push(`Binary file ${name} deleted`);
      } else {
        diffs.push(formatUnifiedDiff(name, '/dev/null', fileContent.toString(), ''));
      }
    } else {
      const stat = fs.lstatSync(fullPath);
      let workingContent: Buffer;

      if (stat.isSymbolicLink()) {
        workingContent = Buffer.from(fs.readlinkSync(fullPath));
      } else {
        workingContent = fs.readFileSync(fullPath);
      }

      const { content: commitFileContent } = readObject(repoRoot, fileSha);

      if (isBinaryFile(commitFileContent) || isBinaryFile(workingContent)) {
        const workingBlobContent = createBlobContent(workingContent);
        const workingSha = hashObject(workingBlobContent);
        if (workingSha !== fileSha) {
          diffs.push(`Binary files a/${name} and b/${name} differ`);
        }
      } else {
        const commitStr = commitFileContent.toString();
        const workingStr = workingContent.toString();
        if (commitStr !== workingStr) {
          diffs.push(formatUnifiedDiff(name, name, commitStr, workingStr));
        }
      }
    }
  }

  return diffs.join('');
}

function diffFileMaps(
  repoRoot: string,
  files1: Map<string, string>,
  files2: Map<string, string>,
  filterPaths: string[]
): string {
  const diffs: string[] = [];
  const allFiles = new Set([...files1.keys(), ...files2.keys()]);

  for (const name of [...allFiles].sort()) {
    if (filterPaths.length > 0 && !matchesPath(name, filterPaths)) {
      continue;
    }

    const sha1 = files1.get(name);
    const sha2 = files2.get(name);

    if (!sha1) {
      // Added in files2
      const { content } = readObject(repoRoot, sha2!);
      if (isBinaryFile(content)) {
        diffs.push(`Binary file ${name} added`);
      } else {
        diffs.push(formatUnifiedDiff('/dev/null', name, '', content.toString()));
      }
    } else if (!sha2) {
      // Deleted in files2
      const { content } = readObject(repoRoot, sha1);
      if (isBinaryFile(content)) {
        diffs.push(`Binary file ${name} deleted`);
      } else {
        diffs.push(formatUnifiedDiff(name, '/dev/null', content.toString(), ''));
      }
    } else if (sha1 !== sha2) {
      // Modified
      const { content: content1 } = readObject(repoRoot, sha1);
      const { content: content2 } = readObject(repoRoot, sha2);

      if (isBinaryFile(content1) || isBinaryFile(content2)) {
        diffs.push(`Binary files a/${name} and b/${name} differ`);
      } else {
        diffs.push(formatUnifiedDiff(name, name, content1.toString(), content2.toString()));
      }
    }
  }

  return diffs.join('');
}

function collectTreeFiles(repoRoot: string, treeSha: string, prefix: string, files: Map<string, string>): void {
  const { content } = readObject(repoRoot, treeSha);
  const entries = parseTreeContent(content);

  for (const entry of entries) {
    const fullName = prefix ? `${prefix}/${entry.name}` : entry.name;

    if (entry.mode === '40000') {
      collectTreeFiles(repoRoot, entry.sha, fullName, files);
    } else {
      files.set(fullName, entry.sha);
    }
  }
}

function matchesPath(name: string, filterPaths: string[]): boolean {
  for (const filter of filterPaths) {
    if (name === filter || name.startsWith(filter + '/')) {
      return true;
    }
  }
  return false;
}

File: minigit-4/py/commands/diff.py
"""diff command - Show changes."""

import sys
import os
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

from utils import find_repo_root, is_binary
from objects import read_object, Commit, Tree, resolve_object, hash_object
from index import Index
from refs import resolve_head
from diff_algo import format_unified_diff


def get_tree_files(tree_sha: str, repo_root: Path) -> dict:
    """Get all files from a tree as {path: sha}."""
    files = {}

    def walk_tree(sha: str, prefix: str):
        try:
            _, data = read_object(sha, repo_root)
            tree = Tree.deserialize(data)
            for entry in tree.entries:
                path = f'{prefix}{entry.name}' if prefix else entry.name
                if entry.mode.startswith('40'):
                    walk_tree(entry.sha, path + '/')
                else:
                    files[path] = entry.sha
        except:
            pass

    walk_tree(tree_sha, '')
    return files


def get_blob_content(sha: str, repo_root: Path) -> bytes:
    """Get content of a blob."""
    _, data = read_object(sha, repo_root)
    return data


def run(args):
    """Show differences."""
    repo_root = find_repo_root()
    if repo_root is None:
        print("fatal: not a minigit repository", file=sys.stderr)
        return 1

    # Parse arguments
    cached = False
    stat_only = False
    paths = []
    commits = []
    separator_seen = False

    i = 0
    while i < len(args):
        arg = args[i]
        if arg == '--':
            separator_seen = True
        elif separator_seen:
            paths.append(arg)
        elif arg in ('--cached', '--staged'):
            cached = True
        elif arg == '--stat':
            stat_only = True
        elif not arg.startswith('-'):
            # Could be a commit or a path
            try:
                sha = resolve_object(arg, repo_root)
                commits.append(sha)
            except:
                paths.append(arg)
        i += 1

    # Determine what to compare
    index = Index.read(repo_root)

    if len(commits) == 2:
        # Compare two commits
        old_sha, new_sha = commits
        _, old_data = read_object(old_sha, repo_root)
        old_commit = Commit.deserialize(old_data)
        _, new_data = read_object(new_sha, repo_root)
        new_commit = Commit.deserialize(new_data)

        old_files = get_tree_files(old_commit.tree_sha, repo_root)
        new_files = get_tree_files(new_commit.tree_sha, repo_root)

        show_diff(old_files, new_files, repo_root, paths, stat_only, from_tree=True, repo_root_for_worktree=None)

    elif len(commits) == 1:
        # Compare commit to working tree
        commit_sha = commits[0]
        _, data = read_object(commit_sha, repo_root)
        commit = Commit.deserialize(data)

        old_files = get_tree_files(commit.tree_sha, repo_root)

        # Get working tree files for all paths in commit tree
        new_files = {}
        for name in old_files.keys():
            file_path = repo_root / name
            if file_path.exists():
                try:
                    content = get_worktree_content(name, repo_root)
                    sha = hash_object(content, 'blob')
                    new_files[name] = sha
                except:
                    pass

        show_diff(old_files, new_files, repo_root, paths, stat_only, from_tree=False, repo_root_for_worktree=repo_root)

    elif cached:
        # Compare HEAD to index (staged changes)
        head_sha = resolve_head(repo_root)
        if head_sha:
            _, data = read_object(head_sha, repo_root)
            commit = Commit.deserialize(data)
            old_files = get_tree_files(commit.tree_sha, repo_root)
        else:
            old_files = {}

        new_files = {name: entry.sha for name, entry in index.entries.items()}

        show_diff(old_files, new_files, repo_root, paths, stat_only, from_tree=True, repo_root_for_worktree=None)

    else:
        # Compare index to working tree (unstaged changes)
        old_files = {name: entry.sha for name, entry in index.entries.items()}
        new_files = get_worktree_files_from_index(repo_root, index)

        show_diff(old_files, new_files, repo_root, paths, stat_only, from_tree=False, repo_root_for_worktree=repo_root)

    return 0


def get_worktree_files_from_index(repo_root: Path, index: Index) -> dict:
    """Get current working tree files for paths in index as {path: sha}."""
    files = {}
    for name in index.entries.keys():
        file_path = repo_root / name
        if file_path.exists():
            try:
                if file_path.is_symlink():
                    content = os.readlink(file_path).encode()
                else:
                    content = file_path.read_bytes()
                sha = hash_object(content, 'blob')
                files[name] = sha
            except:
                pass
    return files


def get_worktree_content(name: str, repo_root: Path) -> bytes:
    """Get content of a file in working tree."""
    file_path = repo_root / name
    if file_path.exists():
        if file_path.is_symlink():
            return os.readlink(file_path).encode()
        else:
            return file_path.read_bytes()
    return b''


def show_diff(old_files: dict, new_files: dict, repo_root: Path,
              paths: list, stat_only: bool, from_tree: bool,
              repo_root_for_worktree: Path = None):
    """Show diff between two file sets."""
    all_files = set(old_files.keys()) | set(new_files.keys())

    if paths:
        # Filter to specific paths
        filtered = set()
        for p in paths:
            for f in all_files:
                if f == p or f.startswith(p + '/'):
                    filtered.add(f)
        all_files = filtered

    for name in sorted(all_files):
        old_sha = old_files.get(name)
        new_sha = new_files.get(name)

        if old_sha == new_sha:
            continue

        # Get content for stat calculation
        if old_sha:
            old_content = get_blob_content(old_sha, repo_root)
        else:
            old_content = b''

        if new_sha:
            if from_tree:
                new_content = get_blob_content(new_sha, repo_root)
            else:
                # Read from working tree
                new_content = get_worktree_content(name, repo_root_for_worktree or repo_root)
        else:
            new_content = b''

        if stat_only:
            # Calculate insertions/deletions
            try:
                old_lines = old_content.decode('utf-8').splitlines()
                new_lines = new_content.decode('utf-8').splitlines()
                insertions = 0
                deletions = 0
                # Simple line count diff
                if len(new_lines) > len(old_lines):
                    insertions = len(new_lines) - len(old_lines)
                elif len(old_lines) > len(new_lines):
                    deletions = len(old_lines) - len(new_lines)
                # Count changed lines
                for i, line in enumerate(old_lines):
                    if i >= len(new_lines) or line != new_lines[i]:
                        deletions += 1
                for i, line in enumerate(new_lines):
                    if i >= len(old_lines) or line != old_lines[i]:
                        insertions += 1
                # Remove duplicates from resize
                if len(new_lines) > len(old_lines):
                    insertions -= len(new_lines) - len(old_lines)
                elif len(old_lines) > len(new_lines):
                    deletions -= len(old_lines) - len(new_lines)
                stat_str = ''
                if insertions > 0:
                    stat_str += '+' * min(insertions, 10)
                if deletions > 0:
                    stat_str += '-' * min(deletions, 10)
                if not stat_str:
                    stat_str = '+-'
                print(f' {name} | {insertions + deletions} {stat_str}')
            except:
                if old_sha is None:
                    print(f' {name} | new file')
                elif new_sha is None:
                    print(f' {name} | deleted')
                else:
                    print(f' {name} | binary')
            continue

        # Check for binary
        if is_binary(old_content) or is_binary(new_content):
            print(f'diff --git a/{name} b/{name}')
            print(f'Binary files a/{name} and b/{name} differ')
            continue

        # Text diff
        try:
            old_text = old_content.decode('utf-8')
            new_text = new_content.decode('utf-8')
        except UnicodeDecodeError:
            print(f'diff --git a/{name} b/{name}')
            print(f'Binary files a/{name} and b/{name} differ')
            continue

        old_lines = old_text.splitlines(keepends=False)
        new_lines = new_text.splitlines(keepends=False)

        diff_output = format_unified_diff(f'a/{name}', f'b/{name}', old_lines, new_lines)

        if diff_output:
            print(f'diff --git a/{name} b/{name}')
            if old_sha is None:
                print('new file mode 100644')
            elif new_sha is None:
                print('deleted file mode 100644')
            print(diff_output)
